File: scala/leet_code/26.scala
/**
 * This is the Solution object. It contains a method to remove duplicates from a sorted array in-place.
 * The method returns the new length of the array after duplicates have been removed.
 *
 * @example
 * val nums = Array(1,1,2)
 * val result = Solution.removeDuplicates(nums)
 * // result: 2
 * // nums: Array(1,2)
 *
 * @note The method modifies the input array by moving the non-duplicate elements to the front of the array.
 */
object Solution {
  /**
   * Removes duplicates from a sorted array in-place.
   *
   * @param nums A sorted array of integers.
   * @return The new length of the array after duplicates have been removed.
   * @throws IllegalArgumentException if the input array is not sorted.
   */
  def removeDuplicates(nums: Array[Int]): Int = {
    // Check if the input array is sorted
    if (nums.length > 1 && nums.zip(nums.tail).exists { case (a, b) => a > b }) {
      throw new IllegalArgumentException("Input array must be sorted")
    }

    // Check if the input array is empty
    if (nums.isEmpty) return 0

    var i = 0
    // Iterate over the array
    for (j <- 1 until nums.length) {
      // If the current element is not equal to the previous element
      if (nums(j) != nums(i)) {
        // Move the non-duplicate element to the front of the array
        i += 1
        nums(i) = nums(j)
      }
    }
    // Return the new length of the array
    i + 1
  }
}

File: go/leet_code/26.go
// Package leet_code provides solutions for LeetCode problems.
//
// This file provides a solution for removing duplicates from a sorted array in-place such that each element appears only once.
package leet_code

// removeDuplicates removes duplicates from a sorted array in-place such that each element appears only once.
//
// It accepts one parameter:
// - nums: a slice of integers representing the sorted array from which duplicates should be removed.
//
// The function returns an integer representing the new length of the array after duplicates have been removed.
//
// The function iterates over the array, replacing the first duplicate element with the current one if the current element is not equal to the previous one.
//
// Error Handling:
// - If the input array is empty, the function returns 0.
//
// Time complexity analysis:
// - Best-case: O(n), when the array has no duplicates.
// - Worst-case: O(n), when all elements in the array are duplicates.
// - Average-case: O(n), as we always have to iterate over each element of the array.
func removeDuplicates(nums []int) int {
	// If the array is empty, return 0
	if len(nums) == 0 {
		return 0
	}

	// i is the index of the first duplicate element
	i := 0

	// Iterate over the array starting from the second element
	for j := 1; j < len(nums); j++ {
		// If the current element is not equal to the previous one
		if nums[i] != nums[j] {
			// Increment the index of the first duplicate element
			i++

			// Replace the first duplicate element with the current one
			nums[i] = nums[j]
		}
	}

	// Return the new length of the array
	return i + 1
}
